rasterio/_warp.pyx

```diff
@@ -487,3 +491,60 @@ def _reproject(
         if dtypes.is_ndarray(source):
             if hdsin != NULL:
                 _gdal.GDALClose(hdsin)
+
+
+def _calculate_default_transform(
+        src_crs, dst_crs, width, height, left, bottom, right, top,
+        resolution=None, densify_pts=21):
+    """Wraps GDAL's algorithm."""
+
+    cdef void *hTransformArg = NULL
+    cdef int npixels = 0
+    cdef int nlines = 0
+    cdef double extent[4]
+    cdef double geotransform[6]
+    cdef void *osr = NULL
+    cdef char *wkt = NULL
+    cdef InMemoryRaster temp = None
+
+    extent[:] = [0.0, 0.0, 0.0, 0.0]
+    geotransform[:] = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+
+    # Make an in-memory raster dataset we can pass to 
+    # GDALCreateGenImgProjTransformer().
+    transform = from_bounds(left, bottom, right, top, width, height)
+    img = np.ones((height, width))
+
+    osr = _base._osr_from_crs(dst_crs)
+    _gdal.OSRExportToWkt(osr, &wkt)
+    _gdal.OSRDestroySpatialReference(osr)
+
+    with InMemoryRaster(
+            img, transform=transform.to_gdal(), crs=src_crs) as temp:
+        hTransformArg = _gdal.GDALCreateGenImgProjTransformer(
+            temp.dataset, NULL, NULL, wkt, 1, 1000.0, 0)
+        if hTransformArg == NULL:
+            raise ValueError("NULL transformer")
+        log.debug("Created transformer")
+
+        if (_gdal.GDALSuggestedWarpOutput2(
+                temp.dataset, _gdal.GDALGenImgProjTransform, hTransformArg,
+                geotransform, &npixels, &nlines, extent, 0)):
```

Comment:
Maybe add a comment that the geotransform, npixels and nlines are expected to be mutated by the function. That's one of the tricker parts for Python programmers to grok about the GDAL C API. 

`npixels` and `nlines` are passed by reference but it's not obvious that geotransform will change. Can you pass arrays by reference is Cython to be explicit? Probably just a comment would be sufficient.